kratos/spaces/ublas_space.h

```diff
@@ -1009,77 +909,109 @@ class UblasSpace
+    /**
+    * @brief Computes the matrix-vector product y = A*x for a sparse matrix A without adding to the result.
+    * @details This function calculates the product of a compressed sparse row (CSR) matrix 'A'
+    * and a vector 'rIn'. The result is stored in 'rOut', overwriting its previous content.
+    * The computation is parallelized over the rows of the matrix.
+    * @param A The input sparse matrix in CSR format.
+    * @param rIn The input vector.
+    * @param rOut The output vector where the result is stored.
+    */
+    static void ParallelProductNoAdd(
+        const MatrixType& rA,
+        const VectorType& rIn,
+        VectorType& rOut
+        )
     {
-        //create partition
+        const unsigned int number_of_rows = rA.size1(); // Total number of rows in matrix A
+
+        // Ensure the output vector has the correct size.
+        if (rOut.size() != number_of_rows) {
+            rOut.resize(number_of_rows, false);
+        }
+
+        // Create partition
         DenseVector<unsigned int> partition;
         unsigned int number_of_threads = omp_get_max_threads();
-        unsigned int number_of_initialized_rows = A.filled1() - 1;
+        unsigned int number_of_initialized_rows = rA.filled1() - 1;
         CreatePartition(number_of_threads, number_of_initialized_rows, partition);
-        //parallel loop
+
+        // Parallel loop
         #pragma omp parallel
         {
-            int thread_id = omp_get_thread_num();
-            int number_of_rows = partition[thread_id + 1] - partition[thread_id];
-            typename compressed_matrix<TDataType>::index_array_type::const_iterator row_iter_begin = A.index1_data().begin() + partition[thread_id];
-            typename compressed_matrix<TDataType>::index_array_type::const_iterator index_2_begin = A.index2_data().begin()+*row_iter_begin;
-            typename compressed_matrix<TDataType>::value_array_type::const_iterator value_begin = A.value_data().begin()+*row_iter_begin;
-            //                  typename VectorType::iterator output_vec_begin = out.begin()+partition[thread_id];
-
-
-            partial_product_no_add(number_of_rows,
-                                   row_iter_begin,
-                                   index_2_begin,
-                                   value_begin,
-                                   in,
-                                   partition[thread_id],
-                                   out
-                                  );
+            const int thread_id = omp_get_thread_num();
+            const int number_of_rows = partition[thread_id + 1] - partition[thread_id];
+            auto row_iter_begin = rA.index1_data().begin() + partition[thread_id];
+            auto index_2_begin = rA.index2_data().begin() + *row_iter_begin;
+            auto value_begin = rA.value_data().begin() + *row_iter_begin;
+
+            PartialProductNoAdd(number_of_rows,
```

Comment:
this implementation is very much legacy of primordial openmp bugs

Reply:
Doyou want to be updated as well?